app/process_proposal.go

```diff
@@ -85,51 +83,42 @@ func (app *App) ProcessProposal(req abci.RequestProcessProposal) abci.ResponsePr
 			continue
 		}
 
-		for _, msg := range tx.GetMsgs() {
-			if sdk.MsgTypeURL(msg) != types.URLMsgPayForBlob {
-				continue
-			}
+		msgs := tx.GetMsgs()
+		// ignore transactions with more than a single sdk.Msg since we do not
+		// yet support arbitrary sdk.Tx
+		if len(msgs) != 1 {
+			continue
+		}
+		msg := msgs[0]
 
-			pfb, ok := msg.(*types.MsgPayForBlob)
-			if !ok {
-				app.Logger().Error("Msg type does not match MsgPayForBlob URL")
-				continue
-			}
+		pfb, ok := msg.(*types.MsgPayForBlob)
+		if !ok {
+			continue
+		}
 
-			if err = pfb.ValidateBasic(); err != nil {
-				logInvalidPropBlockError(app.Logger(), req.Header, "invalid MsgPayForBlob", err)
-				return abci.ResponseProcessProposal{
-					Result: abci.ResponseProcessProposal_REJECT,
-				}
+		if err = pfb.ValidateBasic(); err != nil {
+			logInvalidPropBlockError(app.Logger(), req.Header, "invalid MsgPayForBlob", err)
+			return abci.ResponseProcessProposal{
+				Result: abci.ResponseProcessProposal_REJECT,
 			}
+		}
 
-			commitment, err := inclusion.GetCommit(cacher, dah, int(malleatedTx.ShareIndex), shares.SparseSharesNeeded(uint32(pfb.BlobSize)))
-			if err != nil {
-				logInvalidPropBlockError(app.Logger(), req.Header, "commitment not found", err)
-				return abci.ResponseProcessProposal{
-					Result: abci.ResponseProcessProposal_REJECT,
-				}
+		commitment, err := inclusion.GetMultiCommit(cacher, dah, malleatedTx.ShareIndexes, pfb.BlobSizes)
+		if err != nil {
+			logInvalidPropBlockError(app.Logger(), req.Header, "commitment not found", err)
+			return abci.ResponseProcessProposal{
+				Result: abci.ResponseProcessProposal_REJECT,
 			}
+		}
 
-			if !bytes.Equal(pfb.ShareCommitment, commitment) {
-				logInvalidPropBlock(app.Logger(), req.Header, "found commitment does not match user's")
-				return abci.ResponseProcessProposal{
-					Result: abci.ResponseProcessProposal_REJECT,
-				}
+		if !bytes.Equal(pfb.ShareCommitment, commitment) {
+			logInvalidPropBlock(app.Logger(), req.Header, "found commitment does not match user's")
+			return abci.ResponseProcessProposal{
+				Result: abci.ResponseProcessProposal_REJECT,
 			}
-
-			commitmentCounter++
 		}
 	}
 
-	// compare the number of MPFBs and blobs, if they aren't
-	// identical, then we already know this block is invalid
-	if commitmentCounter != len(req.BlockData.Blobs) {
-		logInvalidPropBlock(app.Logger(), req.Header, "varying number of MsgPayForBlob and blobs in the same block")
-		return abci.ResponseProcessProposal{
-			Result: abci.ResponseProcessProposal_REJECT,
-		}
-	}
```

Comment:
Why are we removing this? We should still be counting the amount of blobs that PFBs are paying for.

Reply:
> If there is an extra PFB, then the starting positions will also be off.

Are you sure? Don't PFBs use compact shares, couldn't you have multiple PFBs in a single share which means the starting position of the blobs would not be offset?

Reply:
ahh, I see. I didn't word that well, by starting positions I meant the starting index of the blob that wraps the PFB.

If there's an extra PFB, then we check that the share commitment is included using the starting position that wraps it. there's no way to fake that without actually including the blob, so we're gucci